### PR TITLE
WIP - complete printing of all node types

### PR DESCRIPTION
**Pull request by mgrip**
Node kinds left:

- [ ] inline
- [ ] magic
- [ ] nowdoc
- [x] yield
- [x] yieldfrom
- [x] sys
- [x] trait
- [x] constant
- [x] classconstant
- [x] retif
- [ ] eval
- [x] exit
- [ ] halt
- [x] clone
- [x] declare
- [x] global
- [x] static
- [x] include
- [x] goto
- [ ] silent
- [x] try
- [x] catch
- [x] throw
- [ ] closure
- [x] new
- [x] label
- [x] traituse
- [x] traitalias
- [x] traitprecedence
- [ ] error

### CODE REVIEW

**Comment by azz** [comment]
Cool stuff. `@` me if you run into any teething issues with the new API!

**Comment by mgrip** [comment]
will do, on first glance looks awesome  - definitely makes sense. thanks @azz!

**Comment by azz** [comment]
🤣 

image

File: tests/traits/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+trait C extends A implements B {
+  public function someOtherTestFunction() {
+    // 4
+  }
+  public function testFunction() {
+    // 5
+  }
+}
+class ImplementingClass {
+
```
Comment (vjeux) [inline tests/traits/__snapshots__/jsfmt.spec.js.snap]: We’ll want to reuse the logic that ensures that there’s at most one empty line in between lines and none at the beginning or end

File: tests/sys/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+echo $test, $other;
+$info = array('coffee', 'brown', 'caffeine');
+list($drink, $color, $power) = $info;
+list(
+  $reallyReallyReallyReallyLongName,
+  $otherReallyReallyReallyLongName,
+  $lastOne
+) = $info;
+print "test print";
+print ("parens test");
```
Comment (vjeux) [inline tests/sys/__snapshots__/jsfmt.spec.js.snap]: This looks weird with a space between the two

File: tests/declare/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+declare(ticks=1);
+$test = 1;
+declare(ticks=1) {
+  $test = 1;
+}
+declare(encoding='ISO-8859-1');
+$test = 1;
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+<?php
+declare(ticks=1):
```
Comment (vjeux) [inline tests/declare/__snapshots__/jsfmt.spec.js.snap]: In javascript we put spaces around =, is it idiomatic not to in php?

File: src/printer.js [hunk 4]
```diff
@@ -613,21 +688,154 @@ function printStatement(node) {
...
+        concat([
+          "static",
+          indent(
+            concat([
+              line,
+              join(
+                concat([",", line]),
+                node.items.map(item => {
+                  // @TODO: hacking this for now. assignments nested inside a static
+                  // declaration doesn't have the operator set, so printing manually
```
Comment (vjeux) [inline src/printer.js]: Would be nice to send issues/prs on the parser implementation. Prettier has found a ton of edge cases on all the parsers we used and having a good relationship with them was critical to make the project work.

**Comment by vjeux** [comment]
This is weird, it passes locally

**Comment by vjeux** [comment]
I reverted https://github.com/prettier/prettier-php/pull/1/commits/90a11a8a1493afad3fbd65a1ca30223eac79667d which is not passing, I'm not sure why as it's passing locally. Let's merge this and the other PR so that more people can contribute at the same time :)

File: tests/declare/__snapshots__/jsfmt.spec.js.snap (hunk 3 above)
Comment (mgrip) [reply tests/declare/__snapshots__/jsfmt.spec.js.snap]: typically yes - this is a bit of an edge case, its a system level directive http://php.net/manual/en/control-structures.declare.php

File: tests/traits/__snapshots__/jsfmt.spec.js.snap (hunk 1 above)
Comment (mgrip) [reply tests/traits/__snapshots__/jsfmt.spec.js.snap]: ya I agree - was thinking we can come back to it. I'll open an issue right now. Do you know where that logic exists in the js printer?

Comment (vjeux) [reply tests/traits/__snapshots__/jsfmt.spec.js.snap]: In printStatementSequence (and in a few other places):

```js
if (util.isNextLineEmpty(text, stmt) && !isLastStatement(stmtPath)) {
      parts.push(hardline);
    }
```
